Remove user.eclass usage from acct-user.eclass and acct-group.eclass

floppym opened the pull request.

mgorny commented on eclass/acct-group.eclass:
```diff
@@ -156,8 +156,23 @@ acct-group_src_install() {
+	fi
+
+	local opts=( --system )
+	[[ -z ${ROOT} ]] || opts+=( --prefix "${ROOT}" )
```
Please use positive logic, "if ROOT is non-empty, then ..." reads much better than "unless ROOT is empty, ...".

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -373,14 +373,17 @@ acct-user_pkg_pretend() {
 	debug-print-function ${FUNCNAME} "${@}"
 
+	# Replace reserved characters in comment
+	ACCT_USER_COMMENT=${DESCRIPTION//[:,=]/;}
```
Missing eclassdoc.

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -429,27 +429,51 @@ acct-user_src_install() {
-	if [[ -n $(egetent passwd "${ACCT_USER_NAME}") ]]; then
-		_ACCT_USER_ALREADY_EXISTS=1
+	if [[ ${EUID} -ne 0 ]]; then
+		einfo "Insufficient privileges to execute ${FUNCNAME[0]}"
```
Use the same message in group and user cases.

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -429,27 +429,51 @@ acct-user_src_install() {
+	fi
+
+	local groups=( ${_ACCT_USER_GROUPS} )
+
```
Remove this empty line.

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -429,27 +429,51 @@ acct-user_src_install() {
+	)
+
+	if [[ ${_ACCT_USER_ID} -ne -1 ]]; then
+		egetent passwd "${_ACCT_USER_ID}" >/dev/null || opts+=( --uid "${_ACCT_USER_ID}" )
```
Use `if ...`.

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -429,27 +429,51 @@ acct-user_src_install() {
-	enewuser ${ACCT_USER_ENFORCE_ID:+-F} -M "${ACCT_USER_NAME}" \
-		"${_ACCT_USER_ID}" "${_ACCT_USER_SHELL}" "${_ACCT_USER_HOME}" \
-		"${_ACCT_USER_GROUPS// /,}"
+	local groups=( ${_ACCT_USER_GROUPS} )
```
Didn't you just do it a few lines earlier?

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -429,27 +429,51 @@ acct-user_src_install() {
-		"${_ACCT_USER_ID}" "${_ACCT_USER_SHELL}" "${_ACCT_USER_HOME}" \
-		"${_ACCT_USER_GROUPS// /,}"
+	local groups=( ${_ACCT_USER_GROUPS} )
+	opts+=( --gid "${groups[0]}" )
```
Any reason not to include it in the initial `opts` set?

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -429,27 +429,51 @@ acct-user_src_install() {
+		local IFS=,
+		echo "$*"
+	}
+	opts+=( --groups "$(_join_comma "${groups[@]:1}")" )
```
```
local aux_groups=${groups[*]:1}
opts+=( --groups "${aux_groups// /,}" )

floppym commented:
Thanks for the review. I believe I have addressed all of your comments.

I also made a few improvements:

- Added || die to usermod calls
- Added special handling of stderr when calling usermod in pkg_postinst. This avoids displaying a warning message to the user about unlocking an account with an empty password.

thesamesam commented on eclass/user.eclass:
```diff
@@ -1,684 +0,0 @@
-# Copyright 1999-2022 Gentoo Authors
-# Distributed under the terms of the GNU General Public License v2
-
-# @ECLASS: user.eclass
```
You'll have to mark it `@DEAD` and give it formal last-rites on the mailing list for 30 days, but I guess this commit was just for CI.

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -426,27 +331,51 @@ acct-user_src_install() {
 	debug-print-function ${FUNCNAME} "${@}"
 
-	# check if user already exists
-	_ACCT_USER_ALREADY_EXISTS=
-	if [[ -n $(egetent passwd "${ACCT_USER_NAME}") ]]; then
-		_ACCT_USER_ALREADY_EXISTS=1
```
You didn't remove the respective eclassdoc. Don't ask me why I've documented it though.

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -426,27 +331,51 @@ acct-user_src_install() {
+
+	elog "Adding user ${ACCT_USER_NAME}"
+	useradd "${opts[@]}" "${ACCT_USER_NAME}" || die
+	_ACCT_USER_ADDED=1
```
You should unset it earlier, so that it doesn't leak from the calling env.

mgorny commented on eclass/acct-user.eclass:
```diff
@@ -469,26 +398,48 @@ acct-user_pkg_preinst() {
+		# empty password. Hide stderr in a text file and display it if usermod
+		# fails.
+		cat "${T}/usermod-error.log" >&2
+		die
```
I suppose some message here would be good given that this line will be without any context.

mgorny commented:
Oh, shi, sorry, I've applied it to look at the code.

floppym commented on eclass/acct-group.eclass:
```diff
@@ -156,8 +156,32 @@ acct-group_src_install() {
+
+	local opts=( --system )
+
+	if [[ -z ${ACCT_GROUP_ENFORCE_ID} ]]; then
```
No, [[ -z ]] is correct here. From the groupadd man page:
```
       -f, --force
           This option causes the command to simply exit with success status
           if the specified group already exists. When used with -g, and the
           specified GID already exists, another (unique) GID is chosen (i.e.
           -g is turned off).
```
When ACCT_GROUP_ENFORCE_ID is blank, we want groupadd to pick another GID instead of failing if the GID already exists.

floppym replied:
I will note that `useradd` does not have an option equivalent to `groupadd --force`. In acct-user.eclass, I added a `egetent passwd "${_ACCT_USER_ID}"` call to check for an existing UID before adding `--uid "${_ACCT_USER_ID}`.

If you prefer for consistency, I could update acct-group.eclass to do a similar check for the GID before adding `--gid ${_ACCT_GROUP_ID}`. I would then drop the `--force` option.

mgorny replied:
Yeah, please do that. I find this whole `--force` option confusing, and I'm going to guess it's not portable anyway.

floppym replied:
Done.

floppym commented on eclass/acct-user.eclass:
```diff
@@ -373,14 +274,18 @@ acct-user_pkg_pretend() {
 acct-user_src_install() {
 	debug-print-function ${FUNCNAME} "${@}"
 
+	# Replace reserved characters in comment
+	: ${ACCT_USER_COMMENT:=${DESCRIPTION}}
+	ACCT_USER_COMMENT=${ACCT_USER_COMMENT//[:,=]/;}
```
Hmm... we actually need to to the character replacement or validation on _ACCT_USER_COMMENT later in the code.

floppym commented:
I have pushed another round of changes.

mgorny reviewed (approved):
LGTM
